docs/design/structs-interfaces.md

```diff
@@ -0,0 +1,251 @@
+
+```go
+r := Resource{
+  Create: func(ctx context.Context, req ResourceCreateRequest, res ResourceCreateResponse) {},
+}
+```
+and
+```go
+r := Resource{
+  Create: myResourceCreate,
+}
+
+func myResourceCreate(ctx context.Context, req ResourceCreateRequest, res ResourceCreateResponse) {}
+```
+
+> The second form is conventional when constructing the `Resource` struct from the existing `helper/schema` framework.
+
+The variety of ways struct methods can be given a value makes the task of static code analysis harder, as it is harder to identify which fragment of code contains the body of the function belonging to `r.Create`. The convention of declaring named functions introduces many long function names to the global namespace. This is a **disadvantage**.
+
+#### Interfaces
+
+Receiver methods fulfilling the method set of an interface are always named the same: for example, in solution 2, an implementation of the `Resource` interface will define a method called `Create`. This is an **advantage** in the case of static analysis, since method names signify intent.
+
+### Documentation
+
+#### Structs
+
+When navigating documentation, struct value functions are not grouped at all. This is a **disadvantage**.
+
+> For example, see the documentation for [`helper/schema#CreateFunc`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/schema#CreateFunc).
+
+#### Interfaces
+
+When navigating documentation, the functions of an interface are grouped together by their namespace. This is an **advantage**.
+
+> For example, see the documentation for [`net/http#Server`](https://pkg.go.dev/net/http#Server).
+
+### Compile-time checks
+
+#### Structs
+
+There is no check at compile time that a struct has non-nil values for every field.
+
+For example, given the following struct:
+
+```go
+type Resource struct {
+  Create func() error
+}
+```
+
+And an instance `r` of that struct type:
+
+```go
+r := Resource{}
+```
+
+A runtime panic will occur due to nil pointer dereference if the SDK calls `r.Create()`. The SDK must check at runtime that struct functions are set appropriately, so developers do not get feedback at compile time if they have omitted a required function. This is a **disadvantage**.
+
+On the other hand, a compile-time error will occur if a provider developer has made a typo and defined, for example, a `Craete` method instead of `Create`. This is an **advantage**.
+
+#### Interfaces
+
+Compile-time assertions can be made in provider code that a struct instance satisfies a particular interface. For example, the following resource must satisfy `framework.Resource` and the compiler will error if any method name has a typo:
+
+```go
+type computeInstanceResource struct {
+  framework.Resource
+}

```

Comment:
This would actually embed a `framework.Resource` type inside of `computeInstanceResource`. It would never cause a compilation error, as long as `framework.Resource` is a type that exists.

```suggestion
type computeInstanceResource struct {
}

var _ framework.Resource = computeInstanceResource{}
```